localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -326,9 +326,39 @@ def on_after_state_load(self):
                                 exc_info=True,
                             )
 
-                # Restore event source listeners
                 for esm in state.event_source_mappings.values():
-                    EventSourceListener.start_listeners_for_asf(esm, self.lambda_service)
+                    if config.LAMBDA_EVENT_SOURCE_MAPPING == "v2":
+                        # Restores event source workers
+                        function_arn = esm.get("FunctionArn")
+
+                        # TODO: How do we know the event source is up?
+                        # A basic poll to see if the mapped Lambda function is active/failed
+                        if not poll_condition(
+                            lambda: get_function_version_from_arn(function_arn).config.state.state
+                            in [State.Active, State.Failed],
+                            timeout=10,
+                        ):
+                            LOG.warning(
+                                "Creating ESM for Lambda that is not in running state: %s",
+                                function_arn,
+                            )
```

Comment:
What would happen if the warning is printed? 10s might be rather short if LocalStack starts polling a new lambda image, but we also do not want to block here for two long. Will the first invocations just fail and be retried?

Reply:
> What would happen if the warning is printed?

An ESM would be created pointing to a Lambda that is still in the `Pending` or `Inactive` state.

> 10s might be rather short if LocalStack starts polling a new lambda image, but we also do not want to block here for two long.

I am skeptical that we would encounter a polling issue here very often since there is a check that happens prior to ESMs being re-created:

https://github.com/localstack/localstack/blob/650483e43779c68f104009aa01b1d54f11c9cd6c/localstack-core/localstack/services/lambda_/provider.py#L292-L300

> Will the first invocations just fail and be retried?

I believe that the function invoke request in the `LambdaSender` (and configured in the `EsmWorker`) would return a `500` since the function is not in an active state. From the `EsmWorker` perspective, I don't think we retry invocations if there is a fail:

https://github.com/localstack/localstack/blob/650483e43779c68f104009aa01b1d54f11c9cd6c/localstack-core/localstack/services/lambda_/event_source_mapping/senders/lambda_sender.py#L45-L50

Reply:
Ah, I see, although an additional 5s won't make much of a difference, pulling the image can take quite some time. Personally, I am in favor of handling that issue later in the ESM worker itself, and retrying there, instead of blocking the on_state_restore call for too long. We could for example set the ESM to `Enabling` until the function is active, or something like that.

Reply:
Agree with the above. Happy to do this in a follow-up PR or should we introduce this more robust function active checker alongside these changes?